core/src/main/scala/cats/data/Chain.scala

```diff
@@ -356,22 +356,111 @@ sealed abstract class Chain[+A] {
 
   /**
    * Groups elements inside this `Chain` according to the `Order`
-   * of the keys produced by the given mapping function.
+   * of the keys produced by the given key function.
+   *
+   * {{{
+   * scala> import scala.collection.immutable.SortedMap
+   * scala> import cats.data.{Chain, NonEmptyChain}
+   * scala> import cats.implicits._
+   * scala> val chain = Chain(12, -2, 3, -5)
+   * scala> val expectedResult = SortedMap(false -> NonEmptyChain(-2, -5), true -> NonEmptyChain(12, 3))
+   * scala> val result = chain.groupBy(_ >= 0)
+   * scala> result === expectedResult
+   * res0: Boolean = true
+   * }}}
    */
-  final def groupBy[B](f: A => B)(implicit B: Order[B]): SortedMap[B, NonEmptyChain[A]] = {
-    implicit val ordering: Ordering[B] = B.toOrdering
-    var m = SortedMap.empty[B, NonEmptyChain[A]]
-    val iter = iterator
+  final def groupBy[K](key: A => K)(implicit K: Order[K]): SortedMap[K, NonEmptyChain[A]] =
```

Comment:
I'm really torn on the renaming.

As for the type parameter, it does not break source compatibility on Scala 2, but Dotty has named type parameters.

The method parameter however is a source-breaking change, even though it is unlikely that someone would give a name to the singleton parameter of a list.

Having said that, I still think we should err on the side of being conservative, so I'd rather avoid the renaming.